File: symmetric/symmetric-core/src/main/java/org/jumpmind/symmetric/service/impl/ConfigurationService.java
/*
 * Licensed to JumpMind Inc under one or more contributor 
 * license agreements.  See the NOTICE file distributed
 * with this work for additional information regarding 
 * copyright ownership.  JumpMind Inc licenses this file
 * to you under the GNU Lesser General Public License (the
 * "License"); you may not use this file except in compliance
 * with the License. 
 * 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, see           
 * <http://www.gnu.org/licenses/>.
 * 
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License. 
 */
package org.jumpmind.symmetric.service.impl;

import java.io.File;
import java.io.InputStreamReader;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;

import org.apache.commons.lang.StringUtils;
import org.jumpmind.db.io.DatabaseIO;
import org.jumpmind.db.model.Database;
import org.jumpmind.db.platform.IDatabasePlatform;
import org.jumpmind.db.sql.AbstractSqlMap;
import org.jumpmind.db.sql.ISqlRowMapper;
import org.jumpmind.db.sql.Row;
import org.jumpmind.db.sql.SqlScript;
import org.jumpmind.db.sql.UniqueKeyException;
import org.jumpmind.symmetric.common.Constants;
import org.jumpmind.symmetric.common.ParameterConstants;
import org.jumpmind.symmetric.db.ISymmetricDialect;
import org.jumpmind.symmetric.model.Channel;
import org.jumpmind.symmetric.model.ChannelMap;
import org.jumpmind.symmetric.model.Node;
import org.jumpmind.symmetric.model.NodeChannel;
import org.jumpmind.symmetric.model.NodeGroup;
import org.jumpmind.symmetric.model.NodeGroupChannelWindow;
import org.jumpmind.symmetric.model.NodeGroupLink;
import org.jumpmind.symmetric.model.NodeGroupLinkAction;
import org.jumpmind.symmetric.model.NodeSecurity;
import org.jumpmind.symmetric.service.IConfigurationService;
import org.jumpmind.symmetric.service.INodeService;
import org.jumpmind.symmetric.service.IParameterService;

/**
 * @see IConfigurationService
 */
public class ConfigurationService extends AbstractService implements IConfigurationService {

    private INodeService nodeService;

    private Map<String, List<NodeChannel>> nodeChannelCache;

    private long nodeChannelCacheTime;

    private List<Channel> defaultChannels;

    public ConfigurationService(IParameterService parameterService, ISymmetricDialect dialect,
            INodeService nodeService) {
        super(parameterService, dialect);
        this.nodeService = nodeService;
        this.defaultChannels = new ArrayList<Channel>();
        this.defaultChannels.add(new Channel(Constants.CHANNEL_CONFIG, 0, 100, 100, true, 0));
        this.defaultChannels.add(new Channel(Constants.CHANNEL_RELOAD, 1, 1, 1, true, 0));
        this.defaultChannels.add(new Channel(Constants.CHANNEL_DEFAULT, 99999, 1000, 100, true, 0));
    }

    public void saveNodeGroupLink(NodeGroupLink link) {
        if (!doesNodeGroupExist(link.getSourceNodeGroupId())) {
            saveNodeGroup(new NodeGroup(link.getSourceNodeGroupId()));
        }

        if (!doesNodeGroupExist(link.getTargetNodeGroupId())) {
            saveNodeGroup(new NodeGroup(link.getTargetNodeGroupId()));
        }

        if (sqlTemplate.update(getSql("updateNodeGroupLinkSql"), link.getDataEventAction().name(),
                link.getSourceNodeGroupId(), link.getTargetNodeGroupId()) == 0) {
            sqlTemplate.update(getSql("insertNodeGroupLinkSql"), link.getDataEventAction().name(),
                    link.getSourceNodeGroupId(), link.getTargetNodeGroupId());
        }
    }

    public boolean doesNodeGroupExist(String nodeGroupId) {
        boolean exists = false;
        List<NodeGroup> groups = getNodeGroups();
        for (NodeGroup nodeGroup : groups) {
            exists |= nodeGroup.getNodeGroupId().equals(nodeGroupId);
        }
        return exists;
    }

    public void saveNodeGroup(NodeGroup group) {
        if (sqlTemplate.update(getSql("updateNodeGroupSql"), group.getDescription(),
                group.getNodeGroupId()) == 0) {
            sqlTemplate.update(getSql("insertNodeGroupSql"), group.getDescription(),
                    group.getNodeGroupId());
        }
    }

    public void deleteNodeGroup(String nodeGroupId) {
        sqlTemplate.update(getSql("deleteNodeGroupSql"), nodeGroupId);
    }

    public void deleteNodeGroupLink(NodeGroupLink link) {
        sqlTemplate.update(getSql("deleteNodeGroupLinkSql"), link.getSourceNodeGroupId(),
                link.getTargetNodeGroupId());
    }

    public List<NodeGroup> getNodeGroups() {
        return sqlTemplate.query(getSql("selectNodeGroupsSql"), new NodeGroupMapper());
    }

    public List<NodeGroupLink> getNodeGroupLinks() {
        return sqlTemplate.query(getSql("groupsLinksSql"), new NodeGroupLinkMapper());
    }

    public List<NodeGroupLink> getNodeGroupLinksFor(String sourceNodeGroupId) {
        return sqlTemplate.query(getSql("groupsLinksForSql"), new NodeGroupLinkMapper(),
                sourceNodeGroupId);
    }

    public NodeGroupLink getNodeGroupLinkFor(String sourceNodeGroupId, String targetNodeGroupId) {
        List<NodeGroupLink> links = getNodeGroupLinksFor(sourceNodeGroupId);
        Iterator<NodeGroupLink> it = links.iterator();
        while (it.hasNext()) {
            NodeGroupLink nodeGroupLink = (NodeGroupLink) it.next();
            if (!nodeGroupLink.getTargetNodeGroupId().equals(targetNodeGroupId)) {
                it.remove();
            }
        }
        return links.size() > 0 ? links.get(0) : null;
    }

    public boolean isChannelInUse(String channelId) {
        return sqlTemplate.queryForInt(getSql("isChannelInUseSql"), channelId) > 0;
    }

    public void saveChannel(Channel channel, boolean reloadChannels) {
        if (0 == sqlTemplate.update(
                getSql("updateChannelSql"),
                new Object[] { channel.getProcessingOrder(), channel.getMaxBatchSize(),
                        channel.getMaxBatchToSend(), channel.getMaxDataToRoute(),
                        channel.isUseOldDataToRoute() ? 1 : 0,
                        channel.isUseRowDataToRoute() ? 1 : 0,
                        channel.isUsePkDataToRoute() ? 1 : 0, channel.isContainsBigLob() ? 1 : 0,
                        channel.isEnabled() ? 1 : 0, channel.getBatchAlgorithm(),
                        channel.getExtractPeriodMillis(), channel.getChannelId() })) {
            sqlTemplate.update(
                    getSql("insertChannelSql"),
                    new Object[] { channel.getChannelId(), channel.getProcessingOrder(),
                            channel.getMaxBatchSize(), channel.getMaxBatchToSend(),
                            channel.getMaxDataToRoute(), channel.isUseOldDataToRoute() ? 1 : 0,
                            channel.isUseRowDataToRoute() ? 1 : 0,
                            channel.isUsePkDataToRoute() ? 1 : 0,
                            channel.isContainsBigLob() ? 1 : 0, channel.isEnabled() ? 1 : 0,
                            channel.getBatchAlgorithm(), channel.getExtractPeriodMillis() });
        }
        if (reloadChannels) {
            reloadChannels();
        }
    }

    public void saveChannel(NodeChannel nodeChannel, boolean reloadChannels) {
        saveChannel(nodeChannel.getChannel(), reloadChannels);
    }

    public void saveNodeChannel(NodeChannel nodeChannel, boolean reloadChannels) {
        saveChannel(nodeChannel.getChannel(), false);
        saveNodeChannelControl(nodeChannel, reloadChannels);
    }

    public void saveNodeChannelControl(NodeChannel nodeChannel, boolean reloadChannels) {
        if (0 == sqlTemplate.update(
                getSql("updateNodeChannelControlSql"),
                new Object[] { nodeChannel.isSuspendEnabled() ? 1 : 0,
                        nodeChannel.isIgnoreEnabled() ? 1 : 0, nodeChannel.getLastExtractTime(),
                        nodeChannel.getNodeId(), nodeChannel.getChannelId() })) {
            sqlTemplate.update(
                    getSql("insertNodeChannelControlSql"),
                    new Object[] { nodeChannel.getNodeId(), nodeChannel.getChannelId(),
                            nodeChannel.isSuspendEnabled() ? 1 : 0,
                            nodeChannel.isIgnoreEnabled() ? 1 : 0,
                            nodeChannel.getLastExtractTime() });
        }
        if (reloadChannels) {
            reloadChannels();
        }
    }

    public void deleteChannel(Channel channel) {
        sqlTemplate.update(getSql("deleteNodeChannelSql"), new Object[] { channel.getChannelId() });
        sqlTemplate.update(getSql("deleteChannelSql"), new Object[] { channel.getChannelId() });
        reloadChannels();
    }

    public NodeChannel getNodeChannel(String channelId, boolean refreshExtractMillis) {
        return getNodeChannel(channelId, nodeService.findIdentityNodeId(), refreshExtractMillis);
    }

    public NodeChannel getNodeChannel(String channelId, String nodeId, boolean refreshExtractMillis) {
        List<NodeChannel> channels = getNodeChannels(nodeId, refreshExtractMillis);
        for (NodeChannel nodeChannel : channels) {
            if (nodeChannel.getChannelId().equals(channelId)) {
                return nodeChannel;
            }
        }
        return null;
    }

    public List<NodeChannel> getNodeChannels(boolean refreshExtractMillis) {
        return getNodeChannels(nodeService.findIdentityNodeId(), refreshExtractMillis);
    }

    public List<NodeChannel> getNodeChannels(final String nodeId, boolean refreshExtractMillis) {
        boolean loaded = false;
        long channelCacheTimeoutInMs = parameterService
                .getLong(ParameterConstants.CACHE_TIMEOUT_CHANNEL_IN_MS);
        List<NodeChannel> nodeChannels = nodeChannelCache != null ? nodeChannelCache.get(nodeId)
                : null;
        if (System.currentTimeMillis() - nodeChannelCacheTime >= channelCacheTimeoutInMs
                || nodeChannels == null) {
            synchronized (this) {
                if (System.currentTimeMillis() - nodeChannelCacheTime >= channelCacheTimeoutInMs
                        || nodeChannelCache == null || nodeChannelCache.get(nodeId) == null
                        || nodeChannels == null) {
                    if (System.currentTimeMillis() - nodeChannelCacheTime >= channelCacheTimeoutInMs
                            || nodeChannelCache == null) {
                        nodeChannelCache = new HashMap<String, List<NodeChannel>>();
                        nodeChannelCacheTime = System.currentTimeMillis();
                    }
                    nodeChannels = sqlTemplate.query(getSql("selectChannelsSql"),
                            new ISqlRowMapper<NodeChannel>() {
                                public NodeChannel mapRow(Row row) {
                                    NodeChannel nodeChannel = new NodeChannel();
                                    nodeChannel.setChannelId(row.getString("channel_id"));
                                    nodeChannel.setNodeId(nodeId);
                                    // note that 2 is intentionally missing
                                    // here.
                                    nodeChannel.setIgnoreEnabled(row.getBoolean("ignore_enabled"));
                                    nodeChannel.setSuspendEnabled(row.getBoolean("suspend_enabled"));
                                    nodeChannel.setProcessingOrder(row.getInt("processing_order"));
                                    nodeChannel.setMaxBatchSize(row.getInt("max_batch_size"));
                                    nodeChannel.setEnabled(row.getBoolean("enabled"));
                                    nodeChannel.setMaxBatchToSend(row.getInt("max_batch_to_send"));
                                    nodeChannel.setMaxDataToRoute(row.getInt("max_data_to_route"));
                                    nodeChannel.setUseOldDataToRoute(row
                                            .getBoolean("use_old_data_to_route"));
                                    nodeChannel.setUseRowDataToRoute(row
                                            .getBoolean("use_row_data_to_route"));
                                    nodeChannel.setUsePkDataToRoute(row
                                            .getBoolean("use_pk_data_to_route"));
                                    nodeChannel.setContainsBigLobs(row
                                            .getBoolean("contains_big_lob"));
                                    nodeChannel.setBatchAlgorithm(row.getString("batch_algorithm"));
                                    nodeChannel.setLastExtractTime(row
                                            .getDateTime("last_extract_time"));
                                    nodeChannel.setExtractPeriodMillis(row
                                            .getLong("extract_period_millis"));
                                    return nodeChannel;
                                };
                            }, nodeId);
                    nodeChannelCache.put(nodeId, nodeChannels);
                    loaded = true;
                }
            }
        }

        if (!loaded && refreshExtractMillis) {
            // need to read last extracted time from database regardless of
            // whether we used the cache or not.
            // locate the nodes in the cache, and update it.
            final Map<String, NodeChannel> nodeChannelsMap = new HashMap<String, NodeChannel>();

            for (NodeChannel nc : nodeChannels) {
                nodeChannelsMap.put(nc.getChannelId(), nc);
            }

            sqlTemplate.query(getSql("selectNodeChannelControlLastExtractTimeSql"),
                    new ISqlRowMapper<Object>() {
                        public Object mapRow(Row row) {
                            String channelId = row.getString("channel_id");
                            Date extractTime = row.getDateTime("last_extract_time");
                            nodeChannelsMap.get(channelId).setLastExtractTime(extractTime);
                            return null;
                        };
                    }, nodeId);

        }

        return nodeChannels;
    }

    public void reloadChannels() {
        synchronized (this) {
            nodeChannelCache = null;
        }
    }

    public NodeGroupLinkAction getDataEventActionByGroupLinkId(String sourceGroupId,
            String targetGroupId) {
        String code = (String) sqlTemplate.queryForObject(getSql("selectDataEventActionsByIdSql"),
                String.class, sourceGroupId, targetGroupId);
        return NodeGroupLinkAction.fromCode(code);
    }

    public void autoConfigDatabase(boolean force) {
        if (parameterService.is(ParameterConstants.AUTO_CONFIGURE_DATABASE) || force) {
            log.info("Initializing SymmetricDS database.");
            symmetricDialect.initTablesAndFunctions();
            autoConfigChannels();
            autoConfigRegistrationServer();
            parameterService.rereadParameters();
            log.info("Done initializing SymmetricDS database.");
        } else {
            log.info("SymmetricDS is not configured to auto-create the database.");
        }
    }

    protected void autoConfigChannels() {
        if (defaultChannels != null) {
            reloadChannels();
            List<NodeChannel> channels = getNodeChannels(false);
            for (Channel defaultChannel : defaultChannels) {
                if (!defaultChannel.isInList(channels)) {
                    log.info("Auto-configuring {} channel.", defaultChannel.getChannelId());
                    saveChannel(defaultChannel, true);
                } else {
                    log.debug("No need to create channel {}.  It already exists.", defaultChannel.getChannelId());
                }
            }
            reloadChannels();
        }
    }

    protected void autoConfigRegistrationServer() {
        Node node = nodeService.findIdentity();

        if (node == null) {
            buildTablesFromDdlUtilXmlIfProvided();
            loadFromScriptIfProvided();
        }

        node = nodeService.findIdentity();

        if (node == null && StringUtils.isBlank(parameterService.getRegistrationUrl())
                && parameterService.is(ParameterConstants.AUTO_INSERT_REG_SVR_IF_NOT_FOUND, false)) {
            log.info("Inserting rows for node, security, identity and group for registration server");
            String nodeGroupId = parameterService.getNodeGroupId();
            String nodeId = parameterService.getExternalId();
            try {
                nodeService.insertNode(nodeId, nodeGroupId, nodeId, nodeId);
            } catch (UniqueKeyException ex) {
                log.warn("Not inserting node row for {} because it already exists", nodeId);
            }
            nodeService.insertNodeIdentity(nodeId);
            node = nodeService.findIdentity();
            node.setSyncUrl(parameterService.getSyncUrl());
            node.setSyncEnabled(true);
            node.setHeartbeatTime(new Date());
            nodeService.updateNode(node);
            nodeService.insertNodeGroup(node.getNodeGroupId(), null);
            NodeSecurity nodeSecurity = nodeService.findNodeSecurity(nodeId, true);
            nodeSecurity.setInitialLoadTime(new Date());
            nodeSecurity.setRegistrationTime(new Date());
            nodeSecurity.setInitialLoadEnabled(false);
            nodeSecurity.setRegistrationEnabled(false);
            nodeService.updateNodeSecurity(nodeSecurity);
        }
    }

    private boolean buildTablesFromDdlUtilXmlIfProvided() {
        boolean loaded = false;
        String xml = parameterService
                .getString(ParameterConstants.AUTO_CONFIGURE_REG_SVR_DDLUTIL_XML);
        if (!StringUtils.isBlank(xml)) {
            File file = new File(xml);
            URL fileUrl = null;
            if (file.isFile()) {
                try {
                    fileUrl = file.toURI().toURL();
                } catch (MalformedURLException e) {
                    throw new RuntimeException(e);
                }
            } else {
                fileUrl = getClass().getResource(xml);
            }

            if (fileUrl != null) {
                try {
                    log.info("Building database schema from: {}", xml);
                    Database database = new DatabaseIO().read(new InputStreamReader(fileUrl
                            .openStream()));
                    IDatabasePlatform platform = symmetricDialect.getPlatform();
                    platform.createDatabase(database, true, true);
                    loaded = true;
                } catch (Exception e) {
                    log.error(e.getMessage(),e);
                }
            }
        }
        return loaded;
    }

    /**
     * Give the end user the option to provide a script that will load a
     * registration server with an initial SymmetricDS setup.
     * 
     * Look first on the file system, then in the classpath for the SQL file.
     * 
     * @return true if the script was executed
     */
    private boolean loadFromScriptIfProvided() {
        boolean loaded = false;
        String sqlScript = parameterService
                .getString(ParameterConstants.AUTO_CONFIGURE_REG_SVR_SQL_SCRIPT);
        if (!StringUtils.isBlank(sqlScript)) {
            File file = new File(sqlScript);
            URL fileUrl = null;
            if (file.isFile()) {
                try {
                    fileUrl = file.toURI().toURL();
                } catch (MalformedURLException e) {
                    throw new RuntimeException(e);
                }
            } else {
                fileUrl = getClass().getResource(sqlScript);
                if (fileUrl == null) {
                    fileUrl = Thread.currentThread().getContextClassLoader().getResource(sqlScript);
                }
            }

            if (fileUrl != null) {
                new SqlScript(fileUrl, symmetricDialect.getPlatform().getSqlTemplate(), true,
                        SqlScript.QUERY_ENDS, getSymmetricDialect().getPlatform()
                                .getSqlScriptReplacementTokens()).execute();
                loaded = true;
            }
        }
        return loaded;
    }

    public List<NodeGroupChannelWindow> getNodeGroupChannelWindows(String nodeGroupId,
            String channelId) {
        return (List<NodeGroupChannelWindow>) sqlTemplate.query(
                getSql("selectNodeGroupChannelWindowSql"), new NodeGroupChannelWindowMapper(),
                nodeGroupId, channelId);
    }

    public ChannelMap getSuspendIgnoreChannelLists(final String nodeId) {
        ChannelMap map = new ChannelMap();
        List<NodeChannel> ncs = getNodeChannels(nodeId, true);

        if (ncs != null) {
            for (NodeChannel nc : ncs) {
                if (nc.isSuspendEnabled()) {
                    map.addSuspendChannels(nc.getChannelId());
                }
                if (nc.isIgnoreEnabled()) {
                    map.addIgnoreChannels(nc.getChannelId());
                }
            }
        }
        return map;
    }

    public Channel getChannel(String channelId) {
        NodeChannel nodeChannel = getNodeChannel(channelId, false);
        if (nodeChannel != null) {
            return nodeChannel.getChannel();
        } else {
            return null;
        }
    }

    public ChannelMap getSuspendIgnoreChannelLists() {
        return getSuspendIgnoreChannelLists(nodeService.findIdentityNodeId());

    }

    @Override
    protected AbstractSqlMap createSqlMap() {
        return new ConfigurationServiceSqlMap(symmetricDialect.getPlatform(),
                createSqlReplacementTokens());
    }

    class NodeGroupChannelWindowMapper implements ISqlRowMapper<NodeGroupChannelWindow> {
        public NodeGroupChannelWindow mapRow(Row row) {
            NodeGroupChannelWindow window = new NodeGroupChannelWindow();
            window.setNodeGroupId(row.getString("node_group_id"));
            window.setChannelId(row.getString("channel_id"));
            window.setStartTime(row.getTime("start_time"));
            window.setEndTime(row.getTime("end_time"));
            window.setEnabled(row.getBoolean("enabled"));
            return window;
        }
    }

    class NodeGroupLinkMapper implements ISqlRowMapper<NodeGroupLink> {
        public NodeGroupLink mapRow(Row row) {
            NodeGroupLink link = new NodeGroupLink();
            link.setSourceNodeGroupId(row.getString("source_node_group_id"));
            link.setTargetNodeGroupId(row.getString("target_node_group_id"));
            link.setDataEventAction(NodeGroupLinkAction.fromCode(row.getString("data_event_action")));
            return link;
        }
    }

    class NodeGroupMapper implements ISqlRowMapper<NodeGroup> {
        public NodeGroup mapRow(Row row) {
            NodeGroup group = new NodeGroup();
            group.setNodeGroupId(row.getString("node_group_id"));
            group.setDescription(row.getString("description"));
            return group;
        }
    }

    public Map<String, String> getRegistrationRedirectMap() {
        return this.sqlTemplate.queryForMap(getSql("getRegistrationRedirectSql"),
                "registrant_external_id", "registration_node_id");
    }

}

File: symmetric-client-clib/src/service/ConfigurationService.c
/**
 * Licensed to JumpMind Inc under one or more contributor
 * license agreements.  See the NOTICE file distributed
 * with this work for additional information regarding
 * copyright ownership.  JumpMind Inc licenses this file
 * to you under the GNU General Public License, version 3.0 (GPLv3)
 * (the "License"); you may not use this file except in compliance
 * with the License.
 *
 * You should have received a copy of the GNU General Public License,
 * version 3.0 (GPLv3) along with this library; if not, see
 * <http://www.gnu.org/licenses/>.
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
#include "service/ConfigurationService.h"
#include "common/Log.h"

static SymChannel * SymConfigurationService_channelMapper(SymRow *row) {
    SymChannel *channel = SymChannel_new(NULL);
    channel->channelId = row->getStringNew(row, "channel_id");
    channel->processingOrder = row->getInt(row, "processing_order");
    channel->maxBatchSize = row->getInt(row, "max_batch_size");
    channel->enabled = row->getBoolean(row, "enabled");
    channel->maxBatchToSend = row->getInt(row, "max_batch_to_send");
    channel->maxDataToRoute = row->getInt(row, "max_data_to_route");
    channel->useOldDataToRoute = row->getBoolean(row, "use_old_data_to_route");
    channel->useRowDataToRoute = row->getBoolean(row, "use_row_data_to_route");
    channel->usePkDataToRoute = row->getBoolean(row, "use_pk_data_to_route");
    channel->containsBigLob = row->getBoolean(row, "contains_big_lob");
    channel->batchAlgorithm = row->getStringNew(row, "batch_algorithm");
    channel->extractPeriodMillis = row->getLong(row, "extract_period_millis");
    channel->dataLoaderType = row->getStringNew(row, "data_loader_type");
    channel->createTime = row->getDate(row, "create_time");
    channel->lastUpdateBy = row->getStringNew(row, "last_update_by");
    channel->lastUpdateTime = row->getDate(row, "last_update_time");
    channel->reloadFlag = row->getBoolean(row, "reload_flag");
    channel->fileSyncFlag = row->getBoolean(row, "file_sync_flag");
    return channel;
}

static SymNodeGroupLink * SymConfigurationService_nodeGroupLinkMapper(SymRow *row) {
    SymNodeGroupLink *link = SymNodeGroupLink_new(NULL);
    link->sourceNodeGroupId = row->getStringNew(row, "source_node_group_id");
    link->targetNodeGroupId = row->getStringNew(row, "target_node_group_id");
    link->dataEventAction = SymNodeGroupLinkAction_fromCode(row->getStringNew(row, "data_event_action"));
    link->syncConfigEnabled = row->getBoolean(row, "sync_config_enabled");
    link->createTime = row->getDate(row, "create_time");
    link->lastUpdateBy = row->getStringNew(row, "last_update_by");
    link->lastUpdateTime = row->getDate(row, "last_update_time");
    return link;
}

SymList * SymConfigurationService_getNodeGroupLinks(SymConfigurationService *this, unsigned short refreshCache) {
    if (refreshCache) {
        this->nodeGroupLinkCacheTime = 0;
    }
    long cacheTimeoutInMs = this->parameterService->getLong(this->parameterService, SYM_PARAMETER_CACHE_TIMEOUT_NODE_GROUP_LINK_IN_MS, 600000);
    if ((time(NULL) - this->nodeGroupLinkCacheTime) * 1000 >= cacheTimeoutInMs || this->nodeGroupLinksCache == NULL) {
        SymSqlTemplate *sqlTemplate = this->platform->getSqlTemplate(this->platform);
        int error;
        this->nodeGroupLinksCache = sqlTemplate->query(sqlTemplate, SYM_SQL_GROUPS_LINKS_SQL, NULL, NULL, &error,
                (void *) &SymConfigurationService_nodeGroupLinkMapper);
        this->nodeGroupLinkCacheTime = time(NULL);
    }

    return this->nodeGroupLinksCache;
}

SymList * SymConfigurationService_getNodeGroupLinksFor(SymConfigurationService *this, char *sourceNodeGroupId, unsigned short refreshCache) {
    SymList *links = SymConfigurationService_getNodeGroupLinks(this, refreshCache);
    SymList *target = SymList_new(NULL);
    SymIterator *iter = links->iterator(links);
    while (iter->hasNext(iter)) {
        SymNodeGroupLink *nodeGroupLink = (SymNodeGroupLink *) iter->next(iter);
        if (SymStringUtils_equals(nodeGroupLink->sourceNodeGroupId, sourceNodeGroupId)) {
            target->add(target, nodeGroupLink);
        }
    }
    iter->destroy(iter);
    return target;
}

SymNodeGroupLink * SymConfigurationService_getNodeGroupLinkFor(SymConfigurationService *this, char *sourceNodeGroupId, char *targetNodeGroupId,
        unsigned short refreshCache) {
    SymList *links = SymConfigurationService_getNodeGroupLinks(this, refreshCache);
    SymNodeGroupLink *link = NULL;
    SymIterator *iter = links->iterator(links);
    while (iter->hasNext(iter)) {
        SymNodeGroupLink *nodeGroupLink = (SymNodeGroupLink *) iter->next(iter);
        if (SymStringUtils_equals(nodeGroupLink->targetNodeGroupId, targetNodeGroupId) &&
                SymStringUtils_equals(nodeGroupLink->sourceNodeGroupId, sourceNodeGroupId)) {
            link = nodeGroupLink;
            break;
        }
    }
    iter->destroy(iter);
    return link;
}

// TODO: cache the channels
SymMap * SymConfigurationService_getChannels(SymConfigurationService *this, unsigned int refreshCache) {
    int error;
    SymSqlTemplate *sqlTemplate = this->platform->getSqlTemplate(this->platform);
    SymList *list = sqlTemplate->query(sqlTemplate, SYM_SQL_SELECT_CHANNEL, NULL, NULL, &error, (void *) SymConfigurationService_channelMapper);
    SymMap *channels = SymMap_new(NULL, list->size);
    SymIterator *iter = list->iterator(list);
    while (iter->hasNext(iter)) {
        SymChannel *channel = (SymChannel *) iter->next(iter);
        channels->put(channels, channel->channelId, channel);
    }
    iter->destroy(iter);
    list->destroy(list);
    return channels;
}

SymList * /*<SymChannel>*/ SymConfigurationService_getFileSyncChannels(SymConfigurationService *this, unsigned int refreshCache) {
    SymMap* channels = this->getChannels(this, refreshCache);
    SymList* list = channels->values(channels);
    SymList* fileSyncChannels = SymList_new(NULL);

    int i;
    for (i = 0; i < list->size; ++i) {
        SymChannel *channel = list->get(list, i);
        if (channel->fileSyncFlag) {
            fileSyncChannels->add(fileSyncChannels, channel);
        }
    }

    channels->destroy(channels);
    return fileSyncChannels;
}

SymChannel * SymConfigurationService_getChannel(SymConfigurationService *this, char *channelId) {
    // TODO refactor to use getNodeChannels like the Java code.
    SymMap *channels = SymConfigurationService_getChannels(this, 0);
    SymChannel *channel = channels->remove(channels, channelId);
    channels->destroyAll(channels, (void *)SymChannel_destroy);
    return channel;
}

void SymConfigurationService_clearCache(SymConfigurationService *this) {
	// TODO
}

void SymConfigurationService_destroy(SymConfigurationService * this) {
    free(this);
}

SymConfigurationService * SymConfigurationService_new(SymConfigurationService *this, SymParameterService *parameterService, SymDatabasePlatform *platform) {
    if (this == NULL) {
        this = (SymConfigurationService *) calloc(1, sizeof(SymConfigurationService));
    }
    this->parameterService = parameterService;
    this->platform = platform;
    this->getChannels = (void *) &SymConfigurationService_getChannels;
    this->getChannel = (void *) &SymConfigurationService_getChannel;
    this->getNodeGroupLinks = (void *) &SymConfigurationService_getNodeGroupLinks;
    this->getNodeGroupLinkFor = (void *) &SymConfigurationService_getNodeGroupLinkFor;
    this->getNodeGroupLinksFor = (void *) &SymConfigurationService_getNodeGroupLinksFor;
    this->getFileSyncChannels = (void *) &SymConfigurationService_getFileSyncChannels;
    this->clearCache = (void *) &SymConfigurationService_clearCache;
    this->destroy = (void *) &SymConfigurationService_destroy;
    return this;
}
